### ISSUE TITLE
Building with `mpi+mpitrampoline` still uses MPICH

### ISSUE THREAD

**Issue by eschnett** [JuliaPackaging/BinaryBuilderBase.jl#245]

**Comment by giordano** [comment]
I think the fundamental issue is that in https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/574320e35258d111e6f28df74017924fcd6e4cc9/src/Prefix.jl#L605-L622 we blindly collect _all_ the dependencies in the manifest.  We have no mechanism at the moment to take some of them out.  With the `platform` keyword argument to the `*Dependency` structs we can remove dependencies from the project on a per-platform base, which is great, but these per-platform relations aren't propagated through (this is also noted with a [big warning box](https://docs.binarybuilder.org/v0.5.3/building/#Platform-dependent-dependencies) in the documentation), so if even if they don't appear at top level of the project, they may still be indirect dependencies of other dependencies.

To try and address this issue, I had a terrible idea: define a new `ExcludeDependency(...; platforms=...)` struct which has the sole purpose of recording that a dependency should be removed from `installed_jlls` from the snippet above, for the given platforms.  So for example we could have something like `ExcludeDependency("OpenMPI_jll"; platforms=filter(p -> p["mpi"] != "openmpi", platforms))` to say "do not even try to install `OpenMPI_jll` in the prefix if `p["mpi"] != "openmpi"`".  This feels like a horrible lack for a deficiency of our packages resolution, but I believe this is the kind of thing we want to do to get this right.  Another, simpler, approach could be to do a setdiff of all the dependencies and only those compatible with the current platform, with the assumption that if we didn't include them in the top-level project we probably don't want them at all, but I suspect there will be corner cases where this can break quite badly (for example if a dependency is listed multiple times, which is not impossible if it has different properties for different platforms).

**Comment by eschnett** [comment]
That sounds like an awkward approach. In principle, I think there shouldn't be a single, overall "set of dependencies", but rather each to-be-installed package specification should have its own set of dependencies.

Another approach would be to mimic the way in which architectures (such as `libgfortran5`) are handled. They need to be consistent across all packages. A setting `mpi+mpitrampoline` would likewise need to be consistent across all installed packages.

To avoid the "installing MPICH and OpenMPI at the same time" problem, we can further require that the settings `mpi+mpich` and `mpi+mpitrampoline` cannot be installed at the same time. That could be done ad-hoc, by calling `build_tarballs` multiple times from the script, or by running the whole `build_tarballs.jl` script multiple times.

**Comment by eschnett** [comment]
Another observation: Using `mpi+mpitrampoline` works for building `ADIOS2_jll`. It only fails for `openPMD_api_jll`, which has `ADIOS2_jll` as dependency. This indicates that the failure mode has to do with other Yggdrasil packages that depend on `mpi+trampoline` as well.

Is there a way to see how the `ADIOS2` dependency is resolved? Does it correctly pick `ADIOS2-mpi+mpitrampoline`? Or might it accidentally pick `ADIOS2-mpi+mpich`?

**Comment by giordano** [comment]
I think @staticfloat can answer that, the artifact resolution happens in Pkg and I don't follow very well the code there.  Also, my feeling is that the `.pkg/platform_augmentation.jl` file isn't actually used, which would be useful to choose the right artifact.

**Comment by staticfloat** [comment]
Alright, so I looked into this, and unfortunately what you're running up against is Pkg's inability for conditional dependencies.  When you build ADIOS2_jll, you [conditionally add dependencies](https://github.com/JuliaPackaging/Yggdrasil/blob/6c186cb0a72a24c16e83bd8d76d9dc871c0c4b8a/platforms/mpi.jl#L62) on a per-platform basis, but unfortunately, when installing things via `Pkg`, we only have a [single `Project.toml`](https://github.com/JuliaBinaryWrappers/ADIOS2_jll.jl/blob/25d3e067fcc4d1fcd61dfa621d852671593a6b50/Project.toml#L6-L22) and every JLL listed there gets installed.

Why isn't this a problem with other per-platform dependencies like the `Xorg_*` JLLs?  Because in those cases, we have mutually exclusive platform mappings.  E.g. it doesn't matter [that `Xorg_libxcb` isn't available on macOS](https://github.com/JuliaPackaging/Yggdrasil/blob/6c186cb0a72a24c16e83bd8d76d9dc871c0c4b8a/X/Xorg_libxcb/build_tarballs.jl#L27) but is installed during builds of GTK3_jll; because there's no artifact mapping for `Xorg_libxcb` on any macOS platforms, so nothing gets installed.

MPICH [has artifacts _without an MPI tag_](https://github.com/JuliaBinaryWrappers/MPICH_jll.jl/blob/e343d691ca6223fb087eaa831fdba7a0925f6248/Artifacts.toml#L319-L328) though, which means that when building for `x86_64-linux-gnu-mpi+trampoline`, it will still install the MPICH files.  I think the easiest way to fix this is to build MPICH to actually have the `mpi+mpich` tag, so that it only gets installed when explicitly requesting that platform.  This is how we can "coordinate" installation across JLLs.

**Comment by eschnett** [comment]
@staticfloat Thanks. Is this something we can do right now, modifying `MPICH_jll`? Or would this break things? Would `MPI.jl` have to choose the `mpi+mpich` version by default when declaring its dependencies?

**Comment by staticfloat** [comment]
Yes, you can do it right now. I don’t THINK you need to do anything complicated; you don’t even need to add a platform augmentation hook I don’t think, as if the host platform lacks one, it should match the only valid platform in the artifacts toml file.

**Comment by giordano** [comment]
I'm a bit concerned though: does it mean that we need to add extra platform tags to all dependencies of such builds?  That doesn't look very sustainable.

**Comment by staticfloat** [comment]
I thought the whole point of `mpitrampoline` was that dependencies depend on _that_ and not on a particular implementation?

**Comment by eschnett** [comment]
Couldn't the platform tags be added automatically?
